crates/rspack_fs/src/native_fs.rs

```diff
@@ -54,14 +70,56 @@ impl WritableFileSystem for NativeFileSystem {
     FileMetadata::try_from(metadata)
   }
 }
+impl From<FileType> for FileMetadata {
+  fn from(value: FileType) -> Self {
+    FileMetadata {
+      is_directory: value == FileType::Directory,
+      is_file: value == FileType::File,
+      is_symlink: false,
+      // yarn pnp don'ts have following info
+      atime_ms: 0,
+      mtime_ms: 0,
+      ctime_ms: 0,
+      size: 0,
+    }
+  }
+}
 
 #[async_trait::async_trait]
 impl ReadableFileSystem for NativeFileSystem {
   fn read(&self, path: &Utf8Path) -> Result<Vec<u8>> {
+    if self.options.enable_pnp {
+      let path = path.as_std_path();
+      let buffer = match VPath::from(path)? {
+        VPath::Zip(info) => self.pnp_lru.read(info.physical_base_path(), info.zip_path),
+        VPath::Virtual(info) => std::fs::read(info.physical_base_path()),
+        VPath::Native(path) => std::fs::read(&path),
+      };
+      return buffer.map_err(Error::from);
+    }
```

Comment:
I wonder if I could/should write a macro in the `pnp` crate that would make this code less boilerplate-y.